### PR TITLE
setup for error handling

### PR DESCRIPTION
**Pull request by brianpmaher**
[Ticket #40](https://www.scrumdo.com/projects/project/compilers/iteration/123104#story_898939)
- Updated `error()` method
- Updated all calls to `error()` with appropriate arguments
- Caught exception

```
brianmaher:~/psrc/uPascal-Compiler (brian-sean/40-refactor-error-messages-and-handling)$ mono mP.exe testProgs/parseError.mp
TOKEN           LINE  COLUMN    LEXEME
PROGRAM            8       0    program
IDENTIFIER         8       8    parseError
SCOLON             8      18    ;
IDENTIFIER        10       0    const
IDENTIFIER        11       2    one
EQUAL             11       6    =
INTEGER_LIT       11       8    1
SCOLON            11       9    ;
IDENTIFIER        12       2    two
EQUAL             12       6    =
INTEGER_LIT       12       8    2
SCOLON            12       9    ;
VAR               14       0    var
IDENTIFIER        15       2    variable
COLON             15      11    :
INTEGER           15      13    integer
SCOLON            15      20    ;
ERROR             16       2    this_
IDENTIFIER        16       7    _issupposed
TO                16      19    to
NOT               16      22    not
IDENTIFIER        16      26    work
COLON             16      31    :
IDENTIFIER        16      33    real
SCOLON            16      37    ;
BEGIN             18       0    begin
WRITELN           20       2    writeln
LPAREN            20      10    (
STRING_LIT        20      11    Variable:
COMMA             20      23    ,
IDENTIFIER        20      25    variable
RPAREN            20      33    )
SCOLON            20      34    ;
END               22       0    end
PERIOD            22       3    .
EOF               23       0    EOF
Matching program
Matching parseError
Matching ;
System.Exception: PARSE ERROR: expected (BEGIN, FUNCTION, PROCEDURE, VAR), but saw "const" of type IDENTIFIER at column 0 and line 10
  at Parser.error (System.Collections.Generic.List`1 expected) [0x00000] in <filename unknown>:0
  at Parser.block () [0x00000] in <filename unknown>:0
  at Parser.program () [0x00000] in <filename unknown>:0
  at Parser.systemGoal () [0x00000] in <filename unknown>:0
  at Parser.Parse () [0x00000] in <filename unknown>:0
  at Driver.Main (System.String[] args) [0x00000] in <filename unknown>:0
```

### CODE REVIEW

File: parser.cs [hunk 1]
```diff
@@ -20,19 +20,30 @@ public class Parser {
...
-            String.Format(
-                Constants.ERROR_PARSER,
-                expected,
-                __lookahead.Lexeme
-            )
-        );
+    private void error(List<string> expected){
+        string exception = "PARSE ERROR: expected ("; // Message to be displayed
+
+        int i = 1,
```
Comment (JesseStephenBrown) [inline parser.cs]: This belongs in the loop, if we are being OCD

**Comment by brianpmaher** [comment]
This is ready for a fresh code-review

**Comment by JesseStephenBrown** [comment]
I did not review each and every replacement call to error. I spoke to John who said it was not a hard requirement to recover from errors either, so we can just report it and end if we so choose. We do need to check if it's a scan error token though in the error handler, because that will have different reporting requirements.

**Comment by brianpmaher** [comment]
It is a requirement though? I can piggy back that to this PR as well.. I figure we continue until we see `end` or `;` and then start scanning again. BUT you do have to traverse up the tree to the right place..

**Comment by seanrogers2657** [comment]
+1

**Comment by JesseStephenBrown** [comment]
+1

**Pull request merged**